### ISSUE TITLE
Proxy HTTP Calls

### ISSUE THREAD

**Issue by awscherb** [segmentio/analytics-kotlin#166]
**Is your feature request related to a problem? Please describe.**
The old Segment Android library provided a clear way to proxy calls via `ConnectionFactory`. This does not appear present in the Kotlin library, and from the API, it does not seem like there is a way to do this. 

**Describe the solution you'd like**
The ability to setup proxying similar to old library during initialization 

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

**Comment by wenxi-zeng** [comment]
hi @awscherb, yes, there is a way to do this in analytics-kotlin to some extends. if your endpoints/path has the same name as the segment one, you could do it with the following:
```
        analytics = Analytics("WRITE KEY", applicationContext) {
            apiHost = "YOUR PROXY"
        }
```
we're currently working on a feature similar to `ConnectionFactory` that will allows developer fully customize the endpoints.

**Comment by ninniuz** [comment]
Hi @wenxi-zeng do you have any updates on this? 
Your suggestion in the previous comment will not work as the `apiHost` is overwritten once the response from `/settings` is applied.

As far as I can see, the Swift SDK is able to provide a `requestFactory` in their Configuration object that is then leveraged by the HTTPClient out of the box to rewrite the final request.

Any ETA on a fix for this issue?

**Comment by wenxi-zeng** [comment]
requestFactory is now available on `1.13.0`. you can now custom the url by create a requestFactory and pass it to configuration
```
        val requestFactory = object : RequestFactory() {
            override fun settings(cdnHost: String, writeKey: String): HttpURLConnection {
                return openConnection("YOUR SETTINGS ENDPOINT")
            }

            override fun upload(apiHost: String): HttpURLConnection {
                return openConnection("YOUR UPLOAD ENDPOINT")
            }

            override fun openConnection(url: String): HttpURLConnection {
                // override only if you want to custom your connection properties
                val requestedURL = URL(url)
                val connection = requestedURL.openConnection() as HttpURLConnection
                connection.connectTimeout = 15_000 // 15s
                connection.readTimeout = 20_1000 // 20s

                connection.setRequestProperty(
                    "User-Agent",
                    "analytics-kotlin/${Constants.LIBRARY_VERSION}"
                )
                connection.doInput = true
                return connection
            }
        }
```
